Check placeholder records and handle validation

ankitrox opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change
<!--
We must be able to understand the design of your change from this description.  The maintainer reviewing this PR may not have worked with this code recently, so please provide as much detail as possible.

Where possible, please also include:
- verification steps to ensure your change has the desired effects and has not introduced any regressions
- any benefits that will be realized
- any alternative implementations or possible drawbacks that you considered
- screenshots or screencasts
-->

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Closes #78 

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->

1. Add the following record in ads.txt editor and save changes. You shall receive the warning "Your ads.txt indicates no authorized advertising sellers."
```
placeholder.example.com, placeholder, DIRECT, placeholder
```

2. Add following record in ads.txt editor. You will get the error "Line 2: Ads.txt contains placeholder record with another records."
```
google.com, 4254, DIRECT
placeholder.example.com, placeholder, DIRECT, placeholder
```

### Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Added - Placeholder record can be added when no authorized sellers or buyers. 


### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @peterwilsoncc


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests pass.

ankitrox committed:
Check placeholder records and handle validation

ankitrox committed:
Unset warning when error

ankitrox committed:
Fix: Tests and phpcs

peterwilsoncc reviewed (commented):
@ankitrox Thanks for the pull request.

Rather than use a static, would it be possible to check if both placeholder and valid records exist within the save function?

https://github.com/10up/ads-txt/blob/a00bdce22ec238acc0d1d599c448bb62af53d2db/inc/save.php#L34-L46

Statics can be difficult to check and while testing the changes with the `test_validate_line` function, I discovered the static was affecting other tests.

jeffpaul commented:
@ankitrox any chance you're able to handle the code review item above?

Sidsector9 committed:
Merge branch 'develop' into fix/78-placeholder-record

ankitrox committed:
Use options for adding placeholder record presence

ankitrox committed:
Resolve conflicts with develop

ankitrox commented:
@peterwilsoncc Thanks for the review :) 

I wanted those couple of variables to be persistent across multiple function calls, but I agree that static may not be the idea ones as far as unit tests are concerned.

I have changed it to use options table in order for them to be persistent. This will also be handy for unit tests.

peterwilsoncc committed:
Merge branch 'develop' into fix/78-placeholder-record

peterwilsoncc commented on inc/save.php:
```diff
@@ -7,6 +7,16 @@
 
 namespace Adstxt;
 
+// Option name for a flag to tell if ads.txt contains placeholder record.
+if ( ! defined( 'ADS_TXT_HAS_PLACEHOLDER_OPTION_NAME' ) ) {
+	define( 'ADS_TXT_HAS_PLACEHOLDER_OPTION_NAME', 'ads_txt_has_placeholder_record' );
+}
+
+// Option name for a flag to store total number of record lines (temporary).
+if ( ! defined( 'ADS_TXT_RECORD_LINES_OPTION_NAME' ) ) {
+	define( 'ADS_TXT_RECORD_LINES_OPTION_NAME', 'ads_txt_record_lines' );
+}
+
```
```suggestion
```

peterwilsoncc commented on inc/save.php:
```diff
@@ -28,6 +38,7 @@ function save() {
 	$sanitized = array();
 	$errors    = array();
+	$warnings  = array();
 	$response  = array();
```
```suggestion
	$response  = array();
	$has_placeholder_record = false;
```

peterwilsoncc commented on inc/save.php:
```diff
@@ -38,8 +49,16 @@ function save() {
 		if ( ! empty( $result['errors'] ) ) {
```
Above, change the `validate_line()` call to

```php
$result      = validate_line( $line, $line_number, $has_placeholder_record );
```

(Sorry, I can't suggest as it's outside the diff)

peterwilsoncc commented on inc/save.php:
```diff
@@ -38,8 +49,16 @@ function save() {
+
+		if ( ! empty( $result['warnings'] ) ) {
+			$warnings = array_merge( $warnings, $result['warnings'] );
+		}
```
```suggestion
		}

		if ( ! empty( $result['is_placeholder_record'] ) ) {
			$has_placeholder_record = true;
		}
```

peterwilsoncc commented on inc/save.php:
```diff
@@ -38,8 +49,16 @@ function save() {
+		}
 	}
 
+	// Delete the temporary options.
+	delete_option( ADS_TXT_HAS_PLACEHOLDER_OPTION_NAME );
+	delete_option( ADS_TXT_RECORD_LINES_OPTION_NAME );
+
```
```suggestion
```

peterwilsoncc commented on inc/save.php:
```diff
@@ -96,8 +116,12 @@ function save() {
  * }
  */
 function validate_line( $line, $line_number ) {
```
```suggestion
function validate_line( $line, $line_number, $has_placeholder_record = false ) {
```

Will need a docblock changes but I can't add it as it's outside the diff.

peterwilsoncc commented on inc/save.php:
```diff
@@ -96,8 +116,12 @@ function save() {
  * }
  */
 function validate_line( $line, $line_number ) {
+	$has_placeholder_record = get_option( ADS_TXT_HAS_PLACEHOLDER_OPTION_NAME, false );
+	$record_lines           = get_option( ADS_TXT_RECORD_LINES_OPTION_NAME, 0 );
```
```suggestion
	static $record_lines           = 0;
	$is_placeholder_record = false;
```

peterwilsoncc commented on inc/save.php:
```diff
@@ -141,37 +165,69 @@ function validate_line( $line, $line_number ) {
-			$pub_id       = trim( $fields[1] );
-			$account_type = trim( $fields[2] );
+			++$record_lines;
+			update_option( ADS_TXT_RECORD_LINES_OPTION_NAME, $record_lines ); // Increment record lines.
```
```suggestion
```

peterwilsoncc commented on inc/save.php:
```diff
@@ -141,37 +165,69 @@ function validate_line( $line, $line_number ) {
-					'value' => $exchange,
+			// If the file contains placeholder record and no placeholder was already present, set variable.
+			if ( $is_placeholder_record && ! $has_placeholder_record ) {
+				update_option( ADS_TXT_HAS_PLACEHOLDER_OPTION_NAME, true, false );
```
```suggestion
```

peterwilsoncc commented on inc/save.php:
```diff
@@ -192,6 +248,7 @@ function validate_line( $line, $line_number ) {
 	return array(
 		'sanitized' => $sanitized,
 		'errors'    => $errors,
+		'warnings'  => $warnings,
```
```suggestion
		'warnings'  => $warnings,
		'is_placeholder_record' => $is_placeholder_record,
```

peterwilsoncc reviewed (changes requested):
I've added quite a few suggestions inline to avoid thrashing the database with option updates. 

To record the line number, I've switch from an option to a static variable. (This may need further thought)

To record the presence of a placeholder, I've added a value to the returned array in `validate_line()`.

While testing, I noticed that the order of lines affects the warnings shown. If I place a valid record above the placeholder record I am shown the warning "Your ads.txt indicates no authorized advertising sellers."

<img width="563" alt="Screen Shot 2023-07-03 at 10 43 52 am" src="https://github.com/10up/ads-txt/assets/519727/c024ab89-22be-4b13-b983-64d57f154870">

If I place the two lines in the reverse order, I also get a message that "Ads.txt contains placeholder record with another records."

<img width="618" alt="Screen Shot 2023-07-03 at 10 46 26 am" src="https://github.com/10up/ads-txt/assets/519727/eecebe70-9c59-4ca7-8fb6-c9efd52730c0">

ankitrox committed:
Remove redundant constants

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

ankitrox committed:
Update inc/save.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

ankitrox committed:
Update inc/save.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

ankitrox committed:
Update inc/save.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

ankitrox committed:
Update inc/save.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

ankitrox committed:
Update inc/save.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

ankitrox committed:
Update inc/save.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

ankitrox committed:
Update inc/save.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

ankitrox committed:
Update inc/save.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

ankitrox committed:
Fix: Error message

ankitrox committed:
Fix: tests

ankitrox committed:
Fix: lint

ankitrox committed:
param docblock

peterwilsoncc commented:
I just tested this and am still seeing the no authorized resellers warning when the file contains both a placeholder and a valid record. 

I did some testing to see if I could come up with an approach that worked but comment lines ended up suppressing the error if the file contained placeholders and comments (which is just as bad, if not worse).

My thoughts are:

* rename `has_placeholder_records` to `has_only_placeholder_records`, in the validate_file function set it to `null`
* if we find a valid line, set the renamed variable to `false`
* if we find a placeholder and the existing value is `null`, set it to true.
* show the warning if the value is either null or true

We'll need to account for:
* comments: no affect on the value
* empty lines: no affect on the value
* invalid records: no affect on the value
* valid record
* placeholders

ankitrox committed:
Merge branch 'develop' into fix/78-placeholder-record

ankitrox committed:
Fix: no warning for valid records

ankitrox commented:
Thank you @peterwilsoncc 

I have added the fix as per Peter's suggestion and it seems to be working fine. Also, added the unit test for the placeholder record.

CC: @jeffpaul

peterwilsoncc committed:
Validate third field is set.

peterwilsoncc committed:
Remove unrequered lint supressions.

peterwilsoncc reviewed (approved):
This looks good to me and tests well. Thank you so much for your patience as we got this working.

I pushed a couple of minor changes:
* 811fabca5eb4910cb4e24288701e4e0711af6235 to make sure the optional tag value was set
* f1a56e501b201458e92e5565a633f5c56719b104 to remove some sniff suppression that were not needed.
